railgun/hack/generators/manifests/main.go

```diff
@@ -0,0 +1,124 @@
+// incompatibility issue automatically.
+//
+// NOTE: When this was first written using controller-gen configuration options
+//       and kubebuilder flags on the API types was attempted but none of these
+//       things could produce the right result.
+//
+//       The controller-gen CLI can't really do it (easily) because the default
+//       behavior of the marshaller in sigs.k8s.io/yaml is to not emit optional
+//       boolean fields to YAML if they're set to false. 😑
+//
+//       It's quite possible there's a better way to do this and if you find one
+//       we encourage you to put in a PR and replace this.
+//
+// -----------------------------------------------------------------------------
+
+func main() {
+	// handle arguments
+	flag.StringVar(&directory, "directory", ".", "which directory to find CRDs in")
+	flag.Parse()
+
+	// find all the YAML CRDs in the provided directory and patch them
+	if err := filepath.Walk(directory, processFile); err != nil {
+		fmt.Fprintf(os.Stderr, "Error processing files in %s: %s\n", directory, err)
+		os.Exit(1)
+	}
+}
+
+func processFile(path string, info os.FileInfo, err error) error {
+	if err != nil {
+		return err
+	}
+
+	if info.IsDir() {
+		return nil
+	}
+
+	// filter out any files that don't identify themselves as YAML files via file extensions
+	if !strings.HasSuffix(info.Name(), ".yaml") && !strings.HasSuffix(info.Name(), ".yml") {
+		return nil
+	}
+
+	// read in the file contents
+	b, err := ioutil.ReadFile(path)
+	if err != nil {
+		return err
+	}
+
+	// filter out any YAML files that don't contain a v1.CustomResourceDefinition
+	if !bytes.Contains(b, []byte(`kind: CustomResourceDefinition`)) || !bytes.Contains(b, []byte(`apiVersion: apiextensions.k8s.io/v1`)) {
+		return nil
+	}
+
+	// don't allow files that contain multiple objects (controller-gen wont be configured to emit these anyway)
+	yamlCRDs := bytes.Split(b, []byte(`\n---\n`))
+	if len(yamlCRDs) > 1 {
+		return fmt.Errorf("file %s contained multiple objects (%d) which this script doesn't yet support", path, len(yamlCRDs))
+	}
+
+	// generate a patch command to ensure "preserveUnknownFields" is set to false
+	stdout, stderr := new(bytes.Buffer), new(bytes.Buffer)
+	patchCMD := exec.Command("kubectl", "patch", "--type", "merge", "--local", "-o", "yaml", "-f", path, "--patch-file", patchfile)
+	patchCMD.Stdout = stdout
+	patchCMD.Stderr = stderr
+
+	// run the patcher and make sure there are no errors
+	if err := patchCMD.Run(); err != nil {
+		return fmt.Errorf("failed to patch crd in %s: %w; STDOUT=(%s) STDERR=(%s)",
+			path, err, strings.TrimSpace(stdout.String()), strings.TrimSpace(stderr.String()),
+		)
+	}
+
+	// controller-gen will add a YAML delimiter in the header, so we match that for consistency
+	patchedCRD := append([]byte("\n---\n"), stdout.Bytes()...)
+
+	// write the patched CRD back to the original file
+	return ioutil.WriteFile(path, patchedCRD, info.Mode().Perm())
```

Comment:
YAGNI

```suggestion
	return nil
```

Reply:
So this is actually needed, the patch wont change the original file in place it only prints it to stdout. I did not find a flag to change this behavior.

Reply:
talked in zoom, good to resolve